client: stop after client disconnect

langmartin opened the pull request:
# stop after client disconnect

Add group configuration, client, and server support for stopping tasks
on disconnected clients (aka "heartyeet").

client:

- [x] add the `stop_after_client_disconnect` key
- [x] persistently track lastHeartbeat, the client local time of the
      last successful heartbeat round trip
- [x] track allocations with `stop_after_client_disconnect` configured
- [x] trigger allocation destroy (which handles cleanup)
- [x] restore heartbeat/killable allocs when allocs are recovered from disk
- [x] on client restart, don't restart allocs that should have been killed

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,149 @@
+	logger hclog.InterceptLogger) *heartbeatStop {
+
+	h := &heartbeatStop{
+		allocs:      make(map[string]time.Duration),
```
The name `allocs` feels a bit ambiguous. This is a map of allocation IDs to... their configured heartbeat interval?

langmartin replied:
fixed

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+// watch is a loop that checks for allocations that should be stopped. It also manages the
+// registration of allocs to be stopped in a single thread.
```
I was going to ask why we didn't push the heartbeatStop down to the individual alloc runners, then I read this. This is a solid "why" comment :+1:

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,149 @@
+
+		now = time.Now()
+		for allocID, d := range h.allocs {
+			if now.After(h.lastOk.Add(d)) {
```
I re-read through the golang docs on [monotonic clocks](https://golang.org/pkg/time/#hdr-Monotonic_Clocks) a couple times, and I'm concerned about the way we persist the last OK to the state store. My understanding is that a monotonic clock value won't have any meaning outside of a single process, so the monotonic clock value is discarded when we serialize it. I _think_ that means that after a restart we have only wall clock time, in which case if a client is restarted on either side of a DST change*, could this result in the client throwing away all its tasks?

(* I've seen the inside of a multi-$B startup who had all their AWS servers set to Pacific Time, so this isn't as far-fetched as one might think! 😆)

tgross replied:
We probably do this sort of thing lots of places, but just retry later and it's all ok. Here it's kind of a one-way trip.

langmartin replied:
It does mean that we won't have a monotonic component on either side of a reboot, the constructors for parsing a timestamp from any representation produce them without a monotonic component. To take it in two pieces: I think that the only problem related specifically to monotonicity is that we could read a timestamp still in `Now`'s future, and I think that's safe. The DST thing is horrifying, so I've just changed it to serialize the unix timestamp. The comparison of that to `Now` should be ok.

tgross replied:
:+1:

tgross commented on client/client.go:
```diff
@@ -1532,6 +1550,10 @@ func (c *Client) registerAndHeartbeat() {
 }
 
+func (c *Client) lastHeartbeat() time.Time {
+	return c.heartbeatStop.lastOk
```
For safe concurrent access, should this be threaded through the state store the way that `setLockOk()` is? (Probably wrapped in a `getLockOk()` method?)

langmartin replied:
fixed, I left a comment about state store error handling which I think is the right decision, but it's a little tricky

tgross replied:
:+1:

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,149 @@
+
+// watch is a loop that checks for allocations that should be stopped. It also manages the
+// registration of allocs to be stopped in a single thread.
+func (h *heartbeatStop) watch() {
```
Can we thread a context down from the client into this or provide a `stop()` method to give us a way to bail out gracefully on shutdown?

langmartin replied:
👍

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,149 @@
+	}
+
+	runner.Destroy()
+	timeout := time.After(5 * time.Second)
```
Is looks to me like `Destroy` doesn't bypass things like `shutdown_delay`. Is this enough time to wait safely? Seems like the destroy will continue without us either way?

tgross commented on jobspec/parse_group.go:
```diff
@@ -56,6 +56,7 @@ func parseGroups(result *api.Job, list *ast.ObjectList) error {
 			"service",
 			"volume",
 			"scaling",
+			"stop_after_client_disconnect",
```
My understanding is this effectively won't be exposed until we do the server-side work? So it should be ok to delay documenting this until that PR?

langmartin replied:
yes, I pulled it out of the api

langmartin commented:
Ok, after some followup, I've removed the stateful last heartbeat handling, and replaced it with a simplified check that, in the case the client crashes but the workload remains running and the client restarts, the client will wait for a server connection grace period and then stop all heartyeet allocs. This avoids the complexity of the state store write on every heartbeat and the state read on the client init path. We believe this to be an unusual failure mode, so keeping it simple until we know we need the additional complexity seems good.

I went ahead an pre-squashed this pr, planning to merge it as two commits to keep the undone state changes available.

tgross commented on client/heartbeatstop.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func (h *heartbeatStop) shouldStopAfter(now time.Time, interval time.Duration) bool {
+	if h.lastOk.IsZero() {
```
should this method also use the `getLastOk()` accessor in the two places we check `lastOk`?

langmartin replied:
good catch, yup!
